go/indexer/services/l1/service.go

```diff
@@ -0,0 +1,386 @@
+	TransactionByHash(context.Context, common.Hash) (*types.Transaction, bool, error)
+}
+
+var (
+	// weiToGwei is the conversion rate from wei to gwei.
+	weiToGwei = new(big.Float).SetFloat64(1e-18)
+)
+
+func uint64ToBytes(i uint64) []byte {
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], i)
+	return buf[:]
+}
+
+func bytesToUint64(b []byte) uint64 {
+	return binary.BigEndian.Uint64(b)
+}
+
+// Merge function to add two uint64 numbers
+func add(existing, new []byte) []byte {
+	return uint64ToBytes(bytesToUint64(existing) + bytesToUint64(new))
+}
+
+func weiToGwei64(wei *big.Int) float64 {
+	gwei := new(big.Float).SetInt(wei)
+	gwei.Mul(gwei, weiToGwei)
+	gwei64, _ := gwei.Float64()
+	return gwei64
+}
+
+// Driver is an interface for indexing deposits from l1.
+type Driver interface {
+	// Name is an identifier used to prefix logs for a particular service.
+	Name() string
+
+	// Metrics returns the subservice telemetry object.
+	Metrics() *metrics.Metrics
+}
+
+type ServiceConfig struct {
+	Context                 context.Context
+	L1Client                *ethclient.Client
+	ChainID                 *big.Int
+	L1StandardBridgeAddress common.Address
+	ConfDepth               uint64
+	MaxHeaderBatchSize      uint64
+	StartBlockNumber        uint64
+	StartBlockHash          string
+	DB                      *db.Database
+}
+
+type Service struct {
+	cfg    ServiceConfig
+	ctx    context.Context
+	cancel func()
+
+	l1BridgeContract *l1bridge.L1StandardBridgeFilterer
+	backend          Backend
+	headerSelector   HeaderSelector
+
+	metrics *metrics.Metrics
+
+	wg sync.WaitGroup
+}
+
+type IndexerStatus struct {
+	Synced  float64           `json:"synced"`
+	Highest db.L1BlockLocator `json:"highest_block"`
+}
+
+func NewService(cfg ServiceConfig) (*Service, error) {
+	ctx, cancel := context.WithCancel(cfg.Context)
+
+	contract, err := l1bridge.NewL1StandardBridgeFilterer(cfg.L1StandardBridgeAddress, cfg.L1Client)
+	if err != nil {
+		return nil, err
+	}
+
+	// Handle restart logic
+
+	log.Info("Creating L1 Indexer")
+
+	chainID, err := cfg.L1Client.ChainID(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.ChainID != nil {
+		if cfg.ChainID.Cmp(chainID) != 0 {
+			return nil, fmt.Errorf("%w: configured with %d and got %d",
+				errWrongChainID, cfg.ChainID, chainID)
+		}
+	} else {
+		cfg.ChainID = chainID
+	}
+
+	confirmedHeaderSelector, err := NewConfirmedHeaderSelector(HeaderSelectorConfig{
+		ConfDepth:    cfg.ConfDepth,
+		MaxBatchSize: cfg.MaxHeaderBatchSize,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		cfg:              cfg,
+		ctx:              ctx,
+		cancel:           cancel,
+		l1BridgeContract: contract,
+		headerSelector:   confirmedHeaderSelector,
+		backend:          cfg.L1Client,
+	}, nil
+}
+
+func (s *Service) Loop(ctx context.Context) {
+	newHeads := make(chan *types.Header, 1000)
+	subscription, err := s.backend.SubscribeNewHead(s.ctx, newHeads)
```

Comment:
I'm not 100% sure this is safe for reorgs

Reply:
Ah i see there is reorg handling in `Update`

Reply:
Would it halt if there was a reorg?

Reply:
No, it would detect a reorg, log the fork block and continue listening for new heads until one connects the last confirmed prev block. The `headerSelector` implements this by calling `HeaderByNumber` to verify the block range between the last confirmed block (our local tip) and the new head. Headers are considered confirmed after a configurable number of blocks defined by the flag `conf-depth` which defaults to `20`.